pkg/telemetry/prometheus/node.go

```diff
@@ -49,36 +49,28 @@ func init() {
 	initRoomStats(nodeID)
 }
 
-func GetUpdatedNodeStats(prev *livekit.NodeStats) (*livekit.NodeStats, error) {
+func GetUpdatedNodeStats(prev *livekit.NodeStats, prevAverage *livekit.NodeStats) (*livekit.NodeStats, bool, error) {
 	loadAvg, err := loadavg.Get()
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 
 	cpuLoad, numCPUs, err := getCPUStats()
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 
 	updatedAt := time.Now().Unix()
-	elapsed := updatedAt - prev.UpdatedAt
+	elapsed := updatedAt - prevAverage.UpdatedAt
+	computeAverage := elapsed > int64(config.StatsUpdateFrequency.Seconds())
 
 	bytesInNow := bytesIn.Load()
 	bytesOutNow := bytesOut.Load()
 	packetsInNow := packetsIn.Load()
 	packetsOutNow := packetsOut.Load()
 	nackTotalNow := nackTotal.Load()
 
-	if bytesInNow == prev.BytesIn &&
-		bytesOutNow == prev.BytesOut &&
-		packetsInNow == prev.PacketsIn &&
-		packetsOutNow == prev.PacketsOut &&
-		nackTotalNow == prev.NackTotal &&
-		elapsed < forceUpdateInterval {
-		return nil, nil
-	}
```

Comment:
I would still leave some form of this in here, i. e. check if the delta is 0 and not update averages and maybe have another config to force update which is more than 10 seconds. Reason being 10 seconds for update and 10 seconds for analytics stats worker could be just line up so that you get the occasional zeros. So, the idea would be, calculate averages, if zero ignore unless force update interval has elapsed. If non-zero, irrespective of time elapsed update. That way, we are letting data drive it and only falling back to forcing an update if nothing has changed for a while.

Reply:
Delta being zero could simply indicate there isn't any activity though.. With the current system, there isn't a way to know when stats update had actually completed.

The issue with returning nil is that we are suppressing updates from fields that didn't rely on the periodic update, for example, liveliness check in Stats.UpdatedAt and CPU/room/participant metrics.

It also felt fragile to rely on separate constants having to be coordinated sufficiently in order for the system to work. So I wanted to avoid that. Since we are using seconds, elapsed time has to be greater than stats update interval (i.e. minimum of 11s) before it'll compute average again. I'm fine with changing the check to include further buffer (i.e. `elapsed > int64(config.StatsUpdateFrequency.Seconds() + 2)`).

Reply:
Yeah, totally agreed on blocking other important updates due to lack of data. Sorry, my intention was not articulated well.

What I am saying is this
- Commit stats like CPU/memory always (at the frequent interval)
- Always compute average here (on every cycle here).
- If the diffs are non-zero, commit them. Some activity has happened and it is valid data.
- If the diffs are zero, check if enough time has elapsed. If so, commit even if it is zero. This condition happens when there a node is hosting no rooms.

That is what I meant by data driven. Always look at the data. If there is change take it. If not, look to force update. But, the force should be given enough slack to not fight with regular updates. Currently, the average computation is purely based on time if I am understanding the code correctly.

Reply:
makes sense, added a short circuit to update more frequently if data had changed.